File: code/erlang/src/ParallelProgramming/others/SpawnTest.java
package com.ericsson.concurrent;

import java.util.concurrent.CountDownLatch;

public class SpawnTest {
    public static void main(String[] args) throws Exception{
        int numberOfThreads = 1000000;
        new SpawnTest().start(numberOfThreads);
    }
    
    CountDownLatch latch = null;
    
	public void start(int numberOfThreads) throws Exception{
		long startTime = System.currentTimeMillis();
		latch = new CountDownLatch(numberOfThreads);
		for (int i = 0; i < numberOfThreads; i++) {
			new Thread(new Task()).start();
		}
		latch.await();
		long stopTime = System.currentTimeMillis();
		System.out.println(numberOfThreads / ((stopTime - startTime) / 1000.0));
	}

    
    public class Task implements Runnable {
		public void run() {
			latch.countDown();
		}
    }
}

File: code/erlang/src/ParallelProgramming/others/SpawnTest.scala
package com.ericsson

import scala.actors._
import scala.actors.Actor._

object SpawnTest {

  class Task extends Actor {
    def act() {
      receive {
        case caller: Actor => caller ! "ok"
      }
    }
  }

  def main(args: Array[String]) {
    val numberOfThreads: Int = 1000000;

    val startTime: Long = System.currentTimeMillis

    val caller = self
    for (i <- 1 to numberOfThreads) {
      val task = new Task()
      task.start()
      task ! caller
    }

    var finishedThreads = 0
    while (finishedThreads < numberOfThreads) {
      receive {
        case "ok" => finishedThreads += 1
      }
    }

    val stopTime: Long = System.currentTimeMillis
    println(numberOfThreads / ((stopTime - startTime) / 1000.0))

  }
}



